datafusion/physical-plan/src/joins/sort_merge_join.rs

```diff
@@ -989,8 +996,21 @@ impl SMJStream {
                 }
             }
             Ordering::Equal => {
-                if matches!(self.join_type, JoinType::LeftSemi) {
+                if matches!(self.join_type, JoinType::LeftSemi) && self.filter.is_some() {
+                    join_streamed = !self
+                        .streamed_batch
+                        .join_filter_matched_idxs
+                        .contains(&(self.streamed_batch.idx as u64))
+                        && !self.streamed_joined;
```

Comment:
If `self.streamed_joined` is false, `join_filter_matched_idxs` always doesn't contain `self.streamed_batch.idx`, so the two conditions are duplicated as they are both true.

If `self.streamed_joined` is true, this `and` check is failed, the another condition doesn't matter.

I'm not sure what this is added to check.